gc/2023/GC-05-16.md

```diff
@@ -33,3 +33,176 @@ Installation is required, see the calendar invite.
+
+JX: The runtime type information is bound to the struct type reference and when we pass the struct ref to that opcode, the runtime can check.
+
+CW: I wouldn’t expect implementations to have all that machinery setup. Maybe the tags are simpler now because you only need enough for the downcasts. I’ll stop speculating.
+
+JX: Because we need to do the downcasts for this, we also need the runtime checking, so we didn’t require additional environment in the runtime.
+
+RH: In SM, we have a link to the full type information from the RTT. We need that for tracing for GC anyway. If we didn’t have tracing, we could perhaps omit it. Theoretically we could load the type information, index into a field, then compare the type.
+
+BT: There would need to be a subtype check there, too, not just an identity check.
+
+RH: Good point
+
+PP: Bottom line is that we require to be implemented
+
+CW: Depends on the implementation strategy of the engine for downcasts.
+
+PP: TL, know anything about it in v8?
+
+TL: Jakob?
+
+JK: Similar to what Ryan has described, we do have the information but the implementation is not optimized for doing fast access, so that would be pretty slow for getting the field offset. Orthogonal, the type issue on the result, when there are reference and non-reference types in there, I’m not sure how we would do that at all. Our existing subtyping expects all ref types. I think if I had to implement a system like this, I would take a very hard look at something based on virtual vTables you can auto-generate that you can generate from the interface information that the compiler sees. Accessors in the vTable to access the fields. My initial reaction.
+
+TL: Was that describing how you would implement it in the engines or how you would fulfill the functionality in userspace.
+
+JK: TypeScript to Wasm compiler.
+
+RH: You’d need performance data to see why to implement it and what it solves. How much performance work to get it to an acceptable level. I could have a built-in that groks the type information, but it would be very slow. Making it fast would be a considerable amount of work, would need to know if its worth it.
+
+JX: Need performance data for the engines to support this implementation?
+
+RH: That would be one thing
+
+CW: Presumably if you are doing the faster for the user code to generate the separate getters and setters, it would probably be faster, though a little annoying. 
+
+RH: I would guess it would be in that case. but I’m sure we could put the type information in that shape so it would be really fast. Right now it’s not, so we’d need a reason to do the work like that. There might be extra metadata costs too.
+
+AR: This is a form of reflection right? It’s about putting reflection into core wasm. As you might guess, I’m not a big fan of that idea, it seems to be something that should be handled at a higher level than the basic asm instruction set.
```

Comment:
@rossberg Sorry for missing this question in the meeting, a quick reply here: 
- I think the proposed opcode `struct.get/set_indirect` is not related to reflection, it's all about getting field index from operand rather than immediate
- The idea of `interface description table` may be something like reflection, but that's not what we want to bring to the core wasm spec, it's just a feasible usage of the proposed opcode

I'll prepare an issue in `WebAssembly/gc` repo and list some benefits as well as impacts of this opcode for further discussion.

BTW, thanks for your professional questions and suggestions in the meeting, which let me realize there are more things to be considered for the type system

Reply:
@rossberg Sorry for the late reply because I didn't notice the update on this thread. Since this PR has been merged, I've created an issue https://github.com/WebAssembly/gc/issues/397 for further discussion